### PR TITLE
Added support for multiple image formats in QuantumCircuit.draw()

### LINKED ISSUE #6448: Feature Request: multi-format file output in QuantumCircuit.draw(..., filename='...')
Qiskit uses PIL for processing image in the Qiskit circuit drawer, which supports several [image file formats](https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.html). However, it always saves in `image/png`:
```python
from qiskit import *

circuit = QuantumCircuit(2)
circuit.h(0)
circuit.cx(0, 1)
circuit.draw('latex', filename='./file.jpg')

!file file.jpg
```
```
file.jpg: PNG image data, 276 x 90, 8-bit/color RGB, non-interlaced
```

It would be great if the drawer notices the `filename` file extension and saves the image in the correct format.

### PR DESCRIPTION
**Pull request by kessler-frost**
### Summary

Fixes #6448. Also adds a warning if the user tries to save the circuit image using `latex` in a format unsupported by PIL and if the user tries to save using `text` but without the ".txt" filename extension.

### Details and comments

Modified the `image.save(filename)` function to allow PIL to automatically determine file format using its filename extension. If the given format is not supported then a warning is displayed to the user.

### CODE REVIEW

File: qiskit/visualization/circuit_visualization.py
Comment (TharrmashasthaPV) [inline qiskit/visualization/circuit_visualization.py:336]: I think it would be better if this check is performed before line 335. That would allow us to catch if any filename with incorrect extensions are given. We could then just raise a warning and not save it. What do you think ? 
Also, I think ```==``` in this line should be ```!=```. And maybe you could use ```filename.endswith(".txt")``` instead of split and check ?

File: qiskit/visualization/circuit_visualization.py
Comment (TharrmashasthaPV) [inline qiskit/visualization/circuit_visualization.py:460]: I think the consensus was just to raise a warning/error and not convert the format. You could check that out in this [comment](https://github.com/Qiskit/qiskit-terra/issues/6448#issuecomment-846436279) . And I guess the same goes to text drawer's output as well.

**Review by TharrmashasthaPV — COMMENTED** [review]
Hi @kessler-frost . Nice work! Thanks for working on this. I just had a few comments.

File: qiskit/visualization/circuit_visualization.py
Comment (kessler-frost) [reply qiskit/visualization/circuit_visualization.py:336]: I should've used `filename.endswith(".txt")`, my bad. About the warning, I was thinking that maybe we should still allow the saving of the file even if the extension is not `.txt` since we are generating a warning and not an error, which is why I kept it after line 335. Do you think it'll be confusing to the user to see a file containing text but without the `.txt` extension? Let me know what you think is best.

File: qiskit/visualization/circuit_visualization.py
Comment (kessler-frost) [reply qiskit/visualization/circuit_visualization.py:460]: Got it. I was just a bit confused as to whether allow saving the file at all or not. I'll make the necessary changes.

File: qiskit/visualization/circuit_visualization.py
Comment (TharrmashasthaPV) [inline qiskit/visualization/circuit_visualization.py:336]: Should we call this 'WARNING' or 'ERROR' in the message? I am a bit inclined towards 'error' since we are not saving any file but no strong opinions there. Same in line 458.

**Comment by TharrmashasthaPV** [comment]
Work nicely done @kessler-frost . This PR looks good to go. Once the blacking is done, I guess the tests too will pass.

File: qiskit/visualization/circuit_visualization.py
Comment (kessler-frost) [reply qiskit/visualization/circuit_visualization.py:336]: Even I think we should call this an ERROR instead of a warning. I'll change that.

**Comment by kessler-frost** [comment]
@TharrmashasthaPV Thanks! I have fixed the linting errors and the tests are now passing too.

**Review by TharrmashasthaPV — APPROVED** [review]
Thanks! @kessler-frost . This LGTM. @1ucian0 any comments?

**Comment by kessler-frost** [comment]
@TharrmashasthaPV I just updated the release notes to show the change of WARNING into ERROR.

**Review by TharrmashasthaPV — APPROVED** [review]

**Comment by 1ucian0** [comment]
`on hold` blocked by #6450

We are almost there with #6450 (snapshot testing for latex drawer). If we merge that one first, we can do better testing for PRs like this one.

**Comment by TharrmashasthaPV** [comment]
Hi @kessler-frost . Very sorry that this went out of radar. Would it be possible for you to add some tests along this PR? A few tests to check if the error messages are returned correctly and some to test if the output file is of the requested format. To test the first part you could provide an unsupported file format as `filename` and check if the error message returned is exactly what is expected. As for the second part I think you could try something along the lines of [`test/python/visualization/test_circuit_drawer.py`](https://github.com/Qiskit/qiskit-terra/blob/main/test/python/visualization/test_circuit_drawer.py)

**Review by TharrmashasthaPV — CHANGES_REQUESTED** [review]
Could you add some tests?

File: qiskit/visualization/circuit_visualization.py
Comment (javabster) [inline qiskit/visualization/circuit_visualization.py:339]: This should raise an error not log a warning. so something like `raise ValueError("error message")`

File: qiskit/visualization/circuit_visualization.py
Comment (javabster) [inline qiskit/visualization/circuit_visualization.py:466]: As above please raise an error instead of logging a warning

**Review by javabster — CHANGES_REQUESTED** [review]

**Comment by kessler-frost** [comment]
@TharrmashasthaPV @javabster I have Implemented the suggestions and added the tests. Please let me know if there's anything more to do.

**Comment by javabster** [comment]
Hi @kessler-frost thanks for your changes! It looks like some of your CI tests are failing because of latex not being installed. If latex isn't installed I think the test should be skipped (you can see how this is done in a similar way with `HAS_MATPLOTLIB` for matplotlib in `test_circuit_drawer.py`)

**Comment by kessler-frost** [comment]
Yeah, my bad. Should've tested that locally too. Will make the changes and update the branch.

**Comment by kessler-frost** [comment]
Sorry for the delay. All the tests have passed and suggestions implemented. The PR is now ready for merge.

**Review by javabster — APPROVED** [review]

File: qiskit/visualization/circuit_visualization.py
Comment (1ucian0) [inline qiskit/visualization/circuit_visualization.py:347]: Qiskit raises `QiskitError` exceptions, I think in this case `VisualizationError` would be okey. 
```suggestion
            raise VisualizationError(
```

File: qiskit/visualization/circuit_visualization.py
Comment (1ucian0) [inline qiskit/visualization/circuit_visualization.py:349]: Error messages are not multiline. I would suggest something shorter, like:
```suggestion
                "ERROR: filename parameter does not use .txt extension."
```

**Review by 1ucian0 — CHANGES_REQUESTED** [review]
Just very small comments. The rest looks good!

File: qiskit/visualization/circuit_visualization.py
Comment (kessler-frost) [reply qiskit/visualization/circuit_visualization.py]: Got it. Updated the file.

File: qiskit/visualization/circuit_visualization.py
Comment (javabster) [inline qiskit/visualization/circuit_visualization.py:471]: I think this error should be changed as well based on @1ucian0's comment :)

Comment (kessler-frost) [reply qiskit/visualization/circuit_visualization.py]: Made the changes.

**Review by javabster — APPROVED** [review]
LGTM 🚀

**Review by 1ucian0 — APPROVED** [review]

**Comment by 1ucian0** [comment]
Thanks!

**Comment by TharrmashasthaPV** [comment]
@1ucian0 @javabster I think we missed auto-merging after the approval : )

**Comment by 1ucian0** [comment]
Ups. Done.

Thanks!

**Pull request merged**
